File: manual/examples/h5py/BasicWriter.py

'''Writes a NeXus HDF5 file using h5py'''

import h5py    # HDF5 support

print "Write a NeXus HDF5 file"
fileName = "prj_test.nexus.hdf5"
timestamp = "2010-10-18T17:17:04-0500"

# prepare the data
f = open('input.dat',  'r')
TWO_COLUMNS = f.read()
f.close()

data = {'mr': [], 'I00': []}
buffer = TWO_COLUMNS.strip().split("\n")
for row in buffer:
    (x, y) = row.split()
    data['mr'].append(float(x))
    data['I00'].append(int(y))

# create the HDF5 NeXus file
f = h5py.File(fileName, "w")
f.attrs['file_name'] = fileName
f.attrs['file_time'] = timestamp
f.attrs['instrument'] = "APS USAXS at 32ID-B"
f.attrs['creator'] = "$Id$"
f.attrs['NeXus_version'] = "4.3.0"
f.attrs['HDF5_Version'] = h5py.version.hdf5_version
f.attrs['h5py_version'] = h5py.version.version

nxentry = f.create_group("entry")
nxentry.attrs["NX_class"] = "NXentry"   # identify NeXus base class

nxdata = nxentry.create_group("mr_scan")
nxdata.attrs["NX_class"] = "NXdata"   # identify NeXus base class

mr = nxdata.create_dataset("mr", data=data['mr'])
mr.attrs['units'] = "degrees"

i00 = nxdata.create_dataset("I00", data=data['I00'])
i00.attrs['units'] = "counts"

# NeXus wants to provide a default plot
# this is the preferred method to describe the axes
i00.attrs['signal'] = "1"    # indicates the Y axis
i00.attrs['axes'] = "mr"     # name "mr" as X axis
mr.attrs['long_name'] = "USAXS mr (degrees)"
i00.attrs['long_name'] = "USAXS I00 (counts)"

f.close()	# be CERTAIN to close the file
print "wrote file:", fileName

File: manual/source/examples/matlab/basic_writer.m
% Writes a NeXus HDF5 file using matlab

disp 'Write a NeXus HDF5 file'
filename = 'prj_test.nexus.hdf5';
timestamp = '2010-10-18T17:17:04-0500';

% read input data
A = load('input.dat');
mr = A(:,1);
I00 = int32(A(:,2));

% clear out old file, if it exists

delete(filename);

% using the simple h5 interface, there is no way to create a group without
% first creating a dataset; creating the dataset creates all intervening
% groups.

% store x
h5create(filename,'/entry/mr_scan/mr',[length(mr)]);
h5write(filename,'/entry/mr_scan/mr',mr);
h5writeatt(filename,'/entry/mr_scan/mr','units','degrees');
h5writeatt(filename,'/entry/mr_scan/mr','long_name','USAXS mr (degrees)');

% store y
h5create(filename,'/entry/mr_scan/I00',[length(I00)],'DataType','int32');
h5write(filename,'/entry/mr_scan/I00',I00);
h5writeatt(filename,'/entry/mr_scan/I00','units','counts');
h5writeatt(filename,'/entry/mr_scan/I00','long_name','USAXS I00 (counts)');

% indicate that we are plotting y vs. x
h5writeatt(filename,'/','default','entry');
h5writeatt(filename,'/entry','default','mr_scan');
h5writeatt(filename,'/entry/mr_scan','signal','I00');
h5writeatt(filename,'/entry/mr_scan','axes','mr_scan');
h5writeatt(filename,'/entry/mr_scan','mr_scan_indices', int32(0));

% add NeXus metadata
h5writeatt(filename,'/','file_name',filename);
h5writeatt(filename,'/','file_time',timestamp);
h5writeatt(filename,'/','instrument','APS USAXS at 32ID-B');
h5writeatt(filename,'/','creator','basic_writer.m');
h5writeatt(filename,'/','NeXus_version','4.3.0');
h5writeatt(filename,'/','HDF5_Version','1.6'); % no 1.8 features used in this example
h5writeatt(filename,'/entry','NX_class','NXentry');
h5writeatt(filename,'/entry/mr_scan','NX_class','NXdata');


h5disp(filename);
